# Guided Capstone Step 4 - Pre-processing and Training Data Development

### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   **Pre-processing and Training Data Development**

5.  Modeling
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

In this step of the guided capstone, you'll revisit some of the steps you completed to solve questions in step 4. Revisiting these steps will help you get additional practice completing this kind of work.

**<font color='teal'> Start by loading the necessary packages, as you've done in the previous steps. Print out your current working directory to confirm that you are in the correct project directory. </font>**

In [120]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import empiricaldist as ed
from scipy.stats import norm
from math import sqrt

%matplotlib inline
os.getcwd()
basepath = os.getcwd()
print(basepath)
lis_dir = os.listdir()
print(lis_dir)


C:\Users\Tiki\Documents\Sprinboard2\Guided_Capstone
['.ipynb_checkpoints', '1582673871_updated_ski_data.csv.zip', 'data', 'figures', 'GuidedCapstone_Steps4and5.ipynb', 'models', 'Notebook_stepfour.ipynb', 'Notebook_stepthree.ipynb', 'Notebook_steptwo.ipynb', 'Ski_Lift_Problem_Statementsheet.pptx', 'step2_output.csv', 'step3_output.csv', 'updated_ski_data.csv', '__MACOSX']


**<font color='teal'>  Load the csv file you created in step 3. Remember, it should be saved inside your data subfolder. Print the first five rows.</font>**

In [121]:
# load csv data
file ='step3_output.csv'
file_name_path = os.path.join(basepath, file)
ski_data= pd.read_csv(file_name_path)
print(ski_data.dtypes)
ski_data.head(5)


Name                  object
state                 object
summit_elev            int64
vertical_drop        float64
trams                  int64
fastEight            float64
fastSixes            float64
fastQuads            float64
quad                 float64
triple               float64
double               float64
surface              float64
total_chairs         float64
Runs                 float64
TerrainParks         float64
LongestRun_mi        float64
SkiableTerrain_ac    float64
Snow Making_ac       float64
daysOpenLastYear     float64
yearsOpen            float64
averageSnowfall      float64
AdultWeekday         float64
AdultWeekend         float64
projectedDaysOpen    float64
NightSkiing_ac       float64
clusters               int64
dtype: object


Name    state  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort   Alaska         3939         2500.0      1        0.0   
1  Eaglecrest Ski Area   Alaska         2600         1540.0      0        0.0   
2     Hilltop Ski Area   Alaska         2090          294.0      0        0.0   
3     Arizona Snowbowl  Arizona        11500         2300.0      0        0.0   
4  Sunrise Park Resort  Arizona        11100         1800.0      0        0.0   

   fastSixes  fastQuads  quad  triple  ...  SkiableTerrain_ac  Snow Making_ac  \
0        0.0        2.0   2.0     0.0  ...             1610.0           113.0   
1        0.0        0.0   0.0     0.0  ...              640.0            60.0   
2        0.0        0.0   0.0     1.0  ...               30.0            30.0   
3        1.0        0.0   2.0     2.0  ...              777.0           104.0   
4        0.0        1.0   2.0     3.0  ...              800.0            80.0   

   daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  AdultWeekend  \
0             150.0       60.0           278.75          65.0          85.0   
1             100.0       44.0           350.00          47.0          53.0   
2             150.0       36.0            69.00          30.0          34.0   
3             122.0       81.0           260.00          89.0          89.0   
4             115.0       49.0           250.00          74.0          78.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              150.0           83.75         0  
1               90.0            0.00         0  
2              152.0           30.00         0  
3              122.0            0.00         1  
4              104.0           80.00         1  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `State`. Add the dummies back to the dataframe and remove the original column for `State`. </font>**

In [122]:
#dfo=df.select_dtypes(include=['object']) # select object type columns
#df = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)

# Create a set of dummy variables from the state variable
popped_state = ski_data['state']
ski_data = pd.concat([ski_data.drop(columns=['state']), pd.get_dummies(ski_data, columns=['state'])], axis=1)
ski_data.head()

Name  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort         3939         2500.0      1        0.0   
1  Eaglecrest Ski Area         2600         1540.0      0        0.0   
2     Hilltop Ski Area         2090          294.0      0        0.0   
3     Arizona Snowbowl        11500         2300.0      0        0.0   
4  Sunrise Park Resort        11100         1800.0      0        0.0   

   fastSixes  fastQuads  quad  triple  double  ...  state_Rhode Island  \
0        0.0        2.0   2.0     0.0     0.0  ...                   0   
1        0.0        0.0   0.0     0.0     4.0  ...                   0   
2        0.0        0.0   0.0     1.0     0.0  ...                   0   
3        1.0        0.0   2.0     2.0     1.0  ...                   0   
4        0.0        1.0   2.0     3.0     1.0  ...                   0   

   state_South Dakota  state_Tennessee  state_Utah  state_Vermont  \
0                   0                0           0              0   
1                   0                0           0              0   
2                   0                0           0              0   
3                   0                0           0              0   
4                   0                0           0              0   

   state_Virginia  state_Washington  state_West Virginia  state_Wisconsin  \
0               0                 0                    0                0   
1               0                 0                    0                0   
2               0                 0                    0                0   
3               0                 0                    0                0   
4               0                 0                    0                0   

   state_Wyoming  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 85 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing, standardize the scale of the features of the dataframe except the name of the resort, which you don't need in the dataframe for modeling so it can be droppped here as well. You should  hold out your response variable(s) so that you can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as your response for scaling and modeling. Later, you will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave, them in the development dataframe. </font>**

In [123]:
#Was there a cell missing here for us to standardize data?

#drop name
name_popped = ski_data.pop('Name')

#import  sklearn
from sklearn import preprocessing

from sklearn.preprocessing import MinMaxScaler 
scaler = MinMaxScaler()


# separate the data and target attributes
ski_data_target = ski_data[['daysOpenLastYear', 'AdultWeekend' ,'AdultWeekday', 'projectedDaysOpen' ]]
ski_data_numeric = ski_data.drop(['AdultWeekend'], axis =1)
#ski_data_numeric2 = ski_data.drop(['daysOpenLastYear', 'AdultWeekend' ,'AdultWeekday', 'projectedDaysOpen' ], axis =1)

print(ski_data.head())
#standardize numeric_ski_Data
list_col = ski_data_numeric.columns
ski_data[list_col] = scaler.fit_transform(ski_data[list_col])
print(ski_data.head())

   summit_elev  vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  \
0         3939         2500.0      1        0.0        0.0        2.0   2.0   
1         2600         1540.0      0        0.0        0.0        0.0   0.0   
2         2090          294.0      0        0.0        0.0        0.0   0.0   
3        11500         2300.0      0        0.0        1.0        0.0   2.0   
4        11100         1800.0      0        0.0        0.0        1.0   2.0   

   triple  double  surface  ...  state_Rhode Island  state_South Dakota  \
0     0.0     0.0      2.0  ...                   0                   0   
1     0.0     4.0      0.0  ...                   0                   0   
2     1.0     0.0      2.0  ...                   0                   0   
3     2.0     1.0      2.0  ...                   0                   0   
4     3.0     1.0      0.0  ...                   0                   0   

   state_Tennessee  state_Utah  state_Vermont  state_Virginia  \
0        

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selectionm import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. You will start by using the adult weekend ticket price as your response variable for modeling.</font>**

In [124]:
from sklearn.model_selection import train_test_split
#ski_data_trial = ski_data
ski_data_trial = ski_data.drop(['AdultWeekend' ,'AdultWeekday', 'projectedDaysOpen' ], axis =1)
#  weekend was highly correlated to weekday price let's not include
# we will also be looking for projected days open so that can't be included

y = ski_data.AdultWeekend
X_train, X_test, y_train, y_test = train_test_split(ski_data_trial , y, test_size=0.25, random_state=0)




Here, you'll start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5 - Modeling

You've cleaned and prepared the datasets, so now it's time to get into the most exciting part of this process: modeling! In this exercise, you will build three different models and compare each model's performance. In the end, you will choose the best model for proving insights to Big Mountain management.  

1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with Training Data Set

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [125]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error, mean_squared_error
lm = linear_model.LinearRegression()
lin_model = lm.fit(X_train,y_train)
y_pred = lm.predict(X_test)

lin_model_coef = {}
for idx, col_name in enumerate(X_train.columns):
    lin_model_coef[col_name] = lm.coef_[0][idx]
    
coef_table =pd.DataFrame(lin_model_coef.items(), columns=['variable', 'coefficent'])


**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

In [126]:

lin_model_Rsq = explained_variance_score(y_test, y_pred)
lin_model_me = mean_absolute_error(y_test, y_pred)
print('The Rsq of the linear model = {} and the Mean Absolute Error = {}'.format(lin_model_Rsq, lin_model_me))


The Rsq of the linear model = 0.40932114164438427 and the MSE = 10.07801735424006


## Review Model Outcomes — Iterate over additional models as needed.

In [127]:
from sklearn import metrics
from sklearn import linear_model
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import GridSearchCV, train_test_split

def test(models, X_train1, X_test1, y_train1, y_test1, iterations = 100):
    results = {}
    for i in models:
        r2_train = []
        r2_test = []
        for j in range(iterations):
            r2_test.append(metrics.r2_score(y_test,
                                            models[i].fit(X_train, 
                                                         y_train).predict(X_test)))
            r2_train.append(metrics.r2_score(y_train, 
                                             models[i].fit(X_train, 
                                                          y_train).predict(X_train)))
        results[i] = [np.mean(r2_train), np.mean(r2_test)]
    return pd.DataFrame(results)
models = {'OLS': linear_model.LinearRegression(),
         'Lasso': linear_model.Lasso(),
         'Ridge': linear_model.Ridge(),}
linmodels1_test = test(models, X_train, X_test, y_train, y_test)
print(linmodels1_test)

        OLS     Lasso     Ridge
0  0.664187  0.350868  0.654552
1  0.386818  0.318856  0.469121


**<font color='teal'> Print the intercept value from the linear model. </font>**

In [128]:
lin_model.intercept_[0]

40.554389787146434

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the `AdultWeekend` resulting value. Also, because you took the time to scale your x values in the training data, you can compare each of the coefficients for the features to determine the feature's importance. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what you are looking for is the magnitude of impact on your response variable.

In [129]:

coef_table['abs_coef']= abs(coef_table.coefficent)
coef_table= coef_table.sort_values(by='abs_coef', ascending= False)
coef_table



variable    coefficent      abs_coef
27        state_Illinois -1.901270e+01  1.901270e+01
51        state_Virginia  1.650596e+01  1.650596e+01
53   state_West Virginia  1.515874e+01  1.515874e+01
36         state_Montana -1.437751e+01  1.437751e+01
39      state_New Jersey  1.339316e+01  1.339316e+01
31        state_Maryland  1.225458e+01  1.225458e+01
26           state_Idaho -1.178580e+01  1.178580e+01
42  state_North Carolina  1.139330e+01  1.139330e+01
43            state_Ohio -1.059301e+01  1.059301e+01
28         state_Indiana -1.019662e+01  1.019662e+01
38   state_New Hampshire  9.937926e+00  9.937926e+00
52      state_Washington -9.829739e+00  9.829739e+00
48       state_Tennessee  9.528158e+00  9.528158e+00
25     state_Connecticut  9.412504e+00  9.412504e+00
37          state_Nevada  8.948052e+00  8.948052e+00
29            state_Iowa -8.594814e+00  8.594814e+00
2                  trams -8.571344e+00  8.571344e+00
47    state_South Dakota -8.139775e+00  8.139775e+00
34       state_Minnesota -7.647605e+00  7.647605e+00
1          vertical_drop  7.531799e+00  7.531799e+00
50         state_Vermont  7.303672e+00  7.303672e+00
44          state_Oregon -7.249501e+00  7.249501e+00
49            state_Utah -6.995579e+00  6.995579e+00
18       averageSnowfall  6.430002e+00  6.430002e+00
9                surface -6.245396e+00  6.245396e+00
10          total_chairs  6.154361e+00  6.154361e+00
5              fastQuads  5.998190e+00  5.998190e+00
55         state_Wyoming -5.518556e+00  5.518556e+00
19        NightSkiing_ac  5.174753e+00  5.174753e+00
32   state_Massachusetts  4.911793e+00  4.911793e+00
45    state_Pennsylvania  4.769445e+00  4.769445e+00
40      state_New Mexico -4.354440e+00  4.354440e+00
6                   quad  4.201556e+00  4.201556e+00
21          state_Alaska -4.039055e+00  4.039055e+00
54       state_Wisconsin  3.825394e+00  3.825394e+00
22         state_Arizona  3.318463e+00  3.318463e+00
7                 triple  3.298603e+00  3.298603e+00
30           state_Maine  2.501706e+00  2.501706e+00
0            summit_elev  2.099495e+00  2.099495e+00
23      state_California -1.923746e+00  1.923746e+00
13         LongestRun_mi  1.720621e+00  1.720621e+00
35        state_Missouri -1.274714e+00  1.274714e+00
33        state_Michigan -1.253912e+00  1.253912e+00
8                 double  1.131318e+00  1.131318e+00
17             yearsOpen  1.096700e+00  1.096700e+00
12          TerrainParks -9.661604e-01  9.661604e-01
24        state_Colorado -9.120905e-01  9.120905e-01
15        Snow Making_ac -6.516103e-01  6.516103e-01
11                  Runs  5.494239e-01  5.494239e-01
41        state_New York  5.363065e-01  5.363065e-01
4              fastSixes -2.465199e-01  2.465199e-01
14     SkiableTerrain_ac  1.265820e-01  1.265820e-01
16      daysOpenLastYear  5.230894e-02  5.230894e-02
20              clusters -3.648643e-02  3.648643e-02
3              fastEight  2.220446e-16  2.220446e-16
46    state_Rhode Island  0.000000e+00  0.000000e+00

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that you care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

#### Model 2

In [135]:
#drop state
ski_data_trial2 = ski_data_trial.loc[:, ~ski_data_trial.columns.str.startswith('state')]

In [140]:

#rerun model excluding state variable
y = ski_data.AdultWeekend
X_train, X_test, y_train, y_test = train_test_split(ski_data_trial2 , y, test_size=0.25, random_state=0)
lm = linear_model.LinearRegression()
lin_model2 = lm.fit(X_train,y_train)
y_pred2 = lm.predict(X_test)

#create Coefficent table
lin_model_coef2 = {}
for idx, col_name in enumerate(X_train.columns):
    lin_model_coef2[col_name] = lm.coef_[0][idx]   
coef_table2 =pd.DataFrame(lin_model_coef2.items(), columns=['variable', 'coefficent'])

#stats_new_model
lin_model2.intercept_[0]
lin_model2_Rsq = explained_variance_score(y_test, y_pred2)
lin_model2_me = mean_absolute_error(y_test, y_pred2)
print('When state dummy variables are removed the Rsq of the linear model = {} and the Mean Absolute Error = {}'.format(lin_model2_Rsq, lin_model2_me))

linmodels2_test = test(models, X_train, X_test, y_train, y_test)

print('R2 for linear models that have state dummy variables')
print(linmodels1_test)

print('R2 for linear models that do not have state dummy variables')
print(linmodels2_test)


When state dummy variables are removed the Rsq of the linear model = 0.40932114164438427 and the MSE = 10.07801735424006
R2 for linear models that have state dummy variables
        OLS     Lasso     Ridge
0  0.664187  0.350868  0.654552
1  0.386818  0.318856  0.469121
R2 for linear models that do not have state dummy variables
        OLS     Lasso     Ridge
0  0.543254  0.350868  0.542223
1  0.513408  0.318856  0.509923


In [141]:
coef_table2['abs_coef']= abs(coef_table2.coefficent)
coef_table2= coef_table2.sort_values(by='abs_coef', ascending= False)
coef_table2

variable  coefficent  abs_coef
2               trams   -9.471792  9.471792
1       vertical_drop    9.351804  9.351804
9             surface   -6.589965  6.589965
5           fastQuads    6.174114  6.174114
10       total_chairs    5.778748  5.778748
6                quad    4.839266  4.839266
19     NightSkiing_ac    3.419033  3.419033
18    averageSnowfall    3.103290  3.103290
7              triple    2.987105  2.987105
13      LongestRun_mi    2.958586  2.958586
14  SkiableTerrain_ac   -2.939699  2.939699
20           clusters   -2.547424  2.547424
17          yearsOpen    2.541847  2.541847
15     Snow Making_ac    2.513896  2.513896
16   daysOpenLastYear    1.721608  1.721608
8              double    1.103706  1.103706
0         summit_elev    1.016092  1.016092
12       TerrainParks   -0.831947  0.831947
4           fastSixes    0.719071  0.719071
11               Runs    0.377899  0.377899
3           fastEight    0.000000  0.000000

**<font color='teal'> When reviewing your new model coefficients, note that `summit_elev` is now in the number two spot. From a managerial perspective, this is also difficult to change and highly correlated with `base_elev` and `vertical_drop`.  This time rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [145]:
ski_data_trial3 = ski_data_trial2.drop(['summit_elev'], axis =1)

#rerun model excluding state variable
y = ski_data.AdultWeekend
X_train, X_test, y_train, y_test = train_test_split(ski_data_trial3 , y, test_size=0.25, random_state=0)
lm = linear_model.LinearRegression()
lin_model3 = lm.fit(X_train,y_train)
y_pred3 = lm.predict(X_test)

#create Coefficent table
lin_model_coef3 = {}
for idx, col_name in enumerate(X_train.columns):
    lin_model_coef3[col_name] = lm.coef_[0][idx]   
coef_table3 =pd.DataFrame(lin_model_coef3.items(), columns=['variable', 'coefficent'])

#stats_new_model
lin_model3.intercept_[0]
lin_model3_Rsq = explained_variance_score(y_test, y_pred3)
lin_model3_me = mean_absolute_error(y_test, y_pred3)
print('When state dummy variables are removed the Rsq of the linear model = {} and the Mean Absolute Error = {}'.format(lin_model3_Rsq, lin_model3_me))

linmodels3_test = test(models, X_train, X_test, y_train, y_test)

print('R2 for linear models that have state dummy variables')
print(linmodels1_test)

print('R2 for linear models that do not have state dummy variables')
print(linmodels2_test)


print('R2 for linear models that do not have state dummy variables and summit elev')
print(linmodels3_test)

When state dummy variables are removed the Rsq of the linear model = 0.5208149819237573 and the MSE = 9.703552758068676
R2 for linear models that have state dummy variables
        OLS     Lasso     Ridge
0  0.664187  0.350868  0.654552
1  0.386818  0.318856  0.469121
R2 for linear models that do not have state dummy variables
        OLS     Lasso     Ridge
0  0.543254  0.350868  0.542223
1  0.513408  0.318856  0.509923
R2 for linear models that do not have state dummy variables and summit elev
        OLS     Lasso     Ridge
0  0.542923  0.350868  0.541873
1  0.519107  0.318856  0.516518


## Identify the Final Model

In [146]:
print('The Rsq of the linear model = {} and the Mean Absolute Error = {}'.format(lin_model_Rsq, lin_model_me))
print('When state dummy variables are removed the Rsq of the linear model = {} and the Mean Absolute Error = {}'.format(lin_model2_Rsq, lin_model2_me))
print('When state dummy variables are removed the Rsq of the linear model = {} and the Mean Absolute Error = {}'.format(lin_model3_Rsq, lin_model3_me))


The Rsq of the linear model = 0.40932114164438427 and the MSE = 10.07801735424006
When state dummy variables are removed the Rsq of the linear model = 0.5155589672085044 and the MSE = 9.735861070356963
When state dummy variables are removed the Rsq of the linear model = 0.5208149819237573 and the MSE = 9.703552758068676


**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell. You will explain your selection during the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.40932114164438427 | 10.078017  |-|
| Model 2. | 0.5155589672085044| 9.735861  |'state'|
| Model 3. | 0.5208149819237573 | 9.703552 |'state','summit_elev','base_elev'|

Model Selection: Model3

In [149]:
#IGNORE
#Create Grid
from sklearn.ensemble import RandomForestRegressor
def evaluate(model, test_features, test_labels):
    predictions = model.predict(test_features)
    errors = abs(predictions - test_labels)
    mape = 100 * np.mean(errors / test_labels)
    accuracy = 100 - mape
    accuracy_value = accuracy.unique()[0]
    avg_error = np.mean(errors)[0]
    print('Model Performance')
    print('Average Error: {} dollars.'.format( avg_error ))
    print('Accuracy = {}%.'.format(accuracy_value))
    return accuracy_value

#Create Random Forest base 
base_model = RandomForestRegressor(random_state=0, n_estimators=20)
base_model.fit(X_train, y_train)
y_pred_for = base_model.predict(X_test)


#evaluate_print(base_accuracy forest)
forest_Rsq = explained_variance_score(y_test, y_pred_for)
forest_me = mean_absolute_error(y_test, y_pred_for)
print('Using Random Forest with 20iterations : The Rsq of the model = {} and the MSE = {}'.format(forest_Rsq, forest_me))
base_accuracy = evaluate(base_model, X_test, y_test)

#evaluate_print(base_accuracy)

#parameters for forest
from sklearn.model_selection import GridSearchCV
param_grid = {"n_estimators": [20, 50, 75, 200, 500], #I prefer less estimators and see how well they do
              "max_depth": [3, None], #3 or no max
              "max_features": [1, 3, 5, 10], #copy pasted
              "min_samples_split": [2, 5, 10], #copy pasted
              "min_samples_leaf": [1, 3, 10], #copy pasted
              "max_features": [1, 3, 5, 10], #copy pasted
              "bootstrap": [True, False] #boost
             }
for_model = RandomForestRegressor(random_state=0)
#grid = GridSearchCV(estimator=for_model, param_grid=param_grid, n_jobs=-1)
#grid.fit(X_train, y_train)
#best_grid = grid.best_estimator_
#grid_accuracy = evaluate(best_grid, X_test, y_test)
#print('Improvement of {}%.'.format( 100 * (grid_accuracy - base_accuracy) / base_accuracy))





Using Random Forest with 20iterations : The Rsq of the model = 0.5054064384966985 and the MSE = 9.733048516647237
Model Performance
Average Error: 9.733048516647237 dollars.
Accuracy = 82.40496300114613%.
